Assets/Mirror/Runtime/NetworkClient.cs

```diff
@@ -100,6 +101,15 @@ protected void OnDataReceived(byte[] data)
             else Debug.LogError("Skipped Data message handling because m_Connection is null.");
         }
 
+        protected void OnDataReceivedNonAlloc(byte[] data, int length)
```

Comment:
why not receive an ArraySegment<byte>  instead?   then transport are not forced to put the message at the beginning of the array.  You can pass the ArraySegment down the line to the NeworkReader too.

Reply:
I don't see any reason to not put the message at the beginning of the array, but i can change that

Reply:
Some transports may want to allocate a large array up front,  and just place messages in random places in that array as they come.  Essentially,  reinventing the heap.

Reply:
I think that they would always place it at the beginning of it, since there is no reason to place them in random places

Reply:
Even if they don't,   passing array and length is how a C programmer would do it.  Do it the C# way,  you have a structure that encapsulates that data and gives you more flexibility in the standard library,  take advantage of it.

Reply:
Okay, i'll change it, but would that change be ok in general?

Reply:
We have to pass the @vis2k  filter,   but I would be perfectly fine changing the receive event to using ArraySegment instead and dropping the byte[] one.
It keeps the code simple and gives the flexibility of eliminating the biggest allocations in Mirror after barriers.
Heck,  I would use it myself in my TcpTransport.

Reply:
Creating NetworkReader per packet allocates much more, but this is easier to fix